### ISSUE TITLE
Make `--format` preserve comments

### ISSUE THREAD

**Issue by byorgey** [swarm-game/swarm#1467]
> I noticed it drops comments, but we can fix that later.

_Originally posted by @xsebek in https://github.com/swarm-game/swarm/issues/1459#issuecomment-1692394374_

**Comment by xsebek** [comment]
I wonder how we would store those.

For formating, it would be nice to have them in the AST, like the SrcLoc.

For parsing though, [it seems easiest](https://www.reddit.com/r/haskell/comments/ni4gpm/how_to_retrieve_comments_using_megaparsec/gz0ipmp/) to put comments in ParserT State monad and pair them afterwards.

@byorgey do you have any experience with round-tripping comments? 😃

**Comment by byorgey** [comment]
> @byorgey do you have any experience with round-tripping comments? 😃

I do not.  It will be a fun learning experience. :smile: 

> For formating, it would be nice to have them in the AST, like the SrcLoc.
> 
> For parsing though, [it seems easiest](https://www.reddit.com/r/haskell/comments/ni4gpm/how_to_retrieve_comments_using_megaparsec/gz0ipmp/) to put comments in ParserT State monad and pair them afterwards.

Agreed... I wonder about a scheme where we save comments out of band via `ParserT State`, and then do a pass at the end where we go through the whole AST and stuff comments into appropriate `Syntax` nodes?

**Comment by xsebek** [comment]
@byorgey that sounds doable, comments should apply to the next `Syntax` node (comparing `SrcLoc`, usually it will be a bind) so one traversal should do it. I assume you mean adding comments to the `Syntax'` data? That is:
```Haskell
data Syntax' ty = Syntax'
  { _sLoc :: SrcLoc
  , _sComment :: Maybe Text -- NEW!
  , _sTerm :: Term' ty
  , _sType :: ty
  }
```

The one parser I know of that tries hard to preserve comments is Spoon for Java, which has quite nice docs:
https://spoon.gforge.inria.fr/comments.html
Java is is a bit different from Swarm, but there is one case (orphan comments) noted in Spoon docs, that we should handle:
```
if (1 + 1 == 2) {
  log "Everything OK in Swarm"
} {
  // will not happen
}
```
In that scenario, we could attach the comment to the implicit `noop`. 🤔

**Comment by byorgey** [comment]
> I assume you mean adding comments to the `Syntax'` data?

Yes, that's exactly what I mean.  It may not even be necessary to do the `State` + second pass thing, maybe we can capture comments directly into `Syntax'` nodes similarly to how we currently capture the `SrcLoc`?  I'm not sure, it would require some playing around.

Yes, we'll probably want a special case for the orphan comment thing.

**Comment by byorgey** [comment]
So far it seems like it has made the most sense to simply collect comments out-of-band as we go, i.e. add a `State (Seq Comment)` effect to the parser.  Then when parsing is done we get an AST along with a sequence of comments (each comment stores its `SrcLoc`).

To actually print these out is the tricky part.  Right now we have an instance 
```
instance PrettyPrec (Syntax' ty) where
  prettyPrec p = prettyPrec p . eraseS
```
which is I guess a good starting point.  We could make a `newtype` which contains a `Syntax` together with a sequence of comments.  The `PrettyPrec` instance for this newtype would look at the `SrcLoc` of the `Syntax` node, and look at the `SrcLoc` of the next comments in the sequence, and decide whether/where to emit the comments.  That's still kind of vague, though, and I'm sure the devil's in the details.

It might be helpful to find some good examples of other language formatters that preserve comments and see how they accomplish it.

**Comment by byorgey** [comment]
Relevant: https://jayconrod.com/posts/129/preserving-comments-when-parsing-and-formatting-code

**Comment by byorgey** [comment]
Progress update: I've started working on this again.  Got `main` merged into the `feature/preserve-comments` branch.  Next steps:
- To apply the technique from the above post, ideally we need to be able to distinguish between comments that occurred by themselves on a line, and comments that occurred after some token on the same line.  How to save this information while parsing?
    - Idea: add a component to the custom state that tracks whether we have seen any tokens on the current line
    - Use a `space` parser which resets this state to false every time it sees a newline
    - Update `lexeme`, `symbol` etc. parsers to set the state to true before beginning
- Update `Syntax'` type to be able to store comments
    - Generalize `Syntax'` so it stores an arbitrary annotation type?
- Create function `populateComments :: Seq Comment -> Syntax -> Syntax` that populates comments into AST using technique from above linked blog post
- Update `readTerm` to apply `populateComments`

**Comment by byorgey** [comment]
We can now parse comments, remembering for each one a source span, whether it was a line or block comment, and whether it was a standalone comment or a suffix on the same line as some other tokens.  For example, this test code:
```
// This function does a cool math thing
def foo : int -> int =     // This is an optional type signature
  \n. n + 1    // add one
end

/* This is a
   block comment which
   spans multiple lines */

def bar : int -> int   // Another type signature, = on the next line
  = \x. foo /* very important to use foo here */ (foo x)   // very cool implementation
end

// And one last thing
```
Produces:
```
Comment {commentSrcLoc = SrcLoc 0 39, commentType = LineComment, commentSituation = StandaloneComment, commentText = " This function does a cool math thing"}
Comment {commentSrcLoc = SrcLoc 67 104, commentType = LineComment, commentSituation = SuffixComment, commentText = " This is an optional type signature"}
Comment {commentSrcLoc = SrcLoc 120 130, commentType = LineComment, commentSituation = SuffixComment, commentText = " add one"}
Comment {commentSrcLoc = SrcLoc 136 198, commentType = BlockComment, commentSituation = StandaloneComment, commentText = " This is a\n   block comment which\n   spans multiple lines "}
Comment {commentSrcLoc = SrcLoc 223 268, commentType = LineComment, commentSituation = SuffixComment, commentText = " Another type signature, = on the next line"}
Comment {commentSrcLoc = SrcLoc 281 317, commentType = BlockComment, commentSituation = SuffixComment, commentText = " very important to use foo here "}
Comment {commentSrcLoc = SrcLoc 328 355, commentType = LineComment, commentSituation = SuffixComment, commentText = " very cool implementation"}
Comment {commentSrcLoc = SrcLoc 361 382, commentType = LineComment, commentSituation = StandaloneComment, commentText = " And one last thing"}
```

**Comment by xsebek** [comment]
Wow, I somehow missed or forgot about this. Many thanks @byorgey, this unblocks #1469.
